OpenRA.Mods.Common/SpriteLoaders/IcnD2Loader.cs

```diff
@@ -0,0 +1,274 @@
+						Data[i++] = pallette[val & 0x0F];
+					}
+			}
+		}
+
+		class IcnD2Frame : ISpriteFrame
+		{
+			public Size Size { get; private set; }
+			public Size FrameSize { get { return Size; } }
+			public float2 Offset { get { return float2.Zero; } }
+			public byte[] Data { get; set; }
+			public bool DisableExportPadding { get { return false; } }
+
+			public IcnD2Frame(int tilesX, int tilesY, int tileIndex, IcnD2Tile[] tiles, ushort[] map)
+			{
+				Size = new Size(SizeX * tilesX, SizeY * tilesY);
+				Data = new byte[Size.Width * Size.Height];
+				var tileOffset = tilesX * tilesY;
+
+				for (var ty = 0; ty < tilesY; ty++)
+					for (var tx = 0; tx < tilesX; tx++)
+					{
+						var j = map[tileIndex * tileOffset + ty * tilesX + tx];
+						for (var y = 0; y < SizeY; y++)
+							for (var x = 0; x < SizeX; x++)
+								Data[(ty * SizeY + y) * Size.Width + (x + tx * SizeX)] = tiles[j].Data[y * SizeX + x];
+					}
+			}
+		}
+
+		bool IsIcnD2(Stream s)
+		{
+			if (s.Length < 0x20)
+				return false;
+
+			var start = s.Position;
+
+			s.Position = 0x18;
+			if (s.ReadByte() != 'S' || s.ReadByte() != 'S' || s.ReadByte() != 'E' || s.ReadByte() != 'T')
+				return false;
+
+			ssetLength = int2.Swap(s.ReadUInt32()) - 8;
+			ssetOffset = 0x18 + 16;
+			if (s.Length < ssetOffset + ssetLength)
+				return false;
+
+			s.Position = ssetOffset + ssetLength;
+			if (s.ReadByte() != 'R' || s.ReadByte() != 'P' || s.ReadByte() != 'A' || s.ReadByte() != 'L')
+				return false;
+
+			rpalLength = int2.Swap(s.ReadUInt32());
+			rpalOffset = ssetOffset + ssetLength + 8;
+			if (s.Length < rpalOffset + rpalLength)
+				return false;
+
+			s.Position = rpalOffset + rpalLength;
+			if (s.ReadByte() != 'R' || s.ReadByte() != 'T' || s.ReadByte() != 'B' || s.ReadByte() != 'L')
+				return false;
+
+			rtblLength = int2.Swap(s.ReadUInt32());
+			rtblOffset = rpalOffset + rpalLength + 8;
+
+			if (s.Length < rtblOffset + rtblLength)
+				return false;
+
+			numTiles = ssetLength / TileSize;
+
+			if (rtblLength < numTiles)
+				return false;
+
+			s.Position = start;
+
+			return true;
+		}
+
+		void ReadTables(Stream s)
+		{
+			var start = s.Position;
+
+			s.Position = rtblOffset;
+			rtbl = StreamExts.ReadBytes(s, (int)rtblLength);
+
+			s.Position = rpalOffset;
+			rpal = new byte[rpalLength / 16][];
+			for (var i = 0; i < rpal.Length; i++)
+				rpal[i] = StreamExts.ReadBytes(s, 16);
+
+			s.Position = start;
+		}
+
+		IcnD2Frame[] ParseFrames(Stream s)
+		{
+			var start = s.Position;
+
+			ReadTables(s);
+
+			var tiles = new IcnD2Tile[numTiles];
+			for (var i = 0; i < tiles.Length; i++)
+			{
+				s.Position = ssetOffset + TileSize * i;
+				tiles[i] = new IcnD2Tile(s, rpal[rtbl[i]]);
+			}
+
+			// first loop. calculate final array size
+			var framesCount = 0;
+			for (var i = 0; i < mapTileSet.Length; i++)
+			{
+				int tilesN;
+				int count = mapTileSet[i].Length;
+				if (framesCount < 210) // structures begins after 210 index
```

Comment:
dune2, as I understand, split sprites that is logicaly should be composed in one building. just because they can operate only blocks 16x16. I think it is easier to use sprites for whole building not sprites for building parts.